python/kotekan/visbuffer.py

```diff
@@ -365,6 +374,160 @@ def __init__(self, filename, mmap=False):
+            "freq": freq,
+            "input": input_,
+            "prod": prod,
+            "ev": list(range(nev)),
+        }
+
+        if stack is not None:
+            if not isinstance(stack, list):
+                raise ValueError("Incorrect format for stack axis")
+            index_map["stack"] = stack
+
+        # Calculate the structural metadata
+        ninput = len(input_)
+        nstack = len(stack) if stack is not None else len(prod)
+        nfreq = len(freq)
+        ntime = len(time)
+
+        msize = ctypes.sizeof(VisMetadata)
+        dsize = VisBuffer._calculate_layout(ninput, nstack, nev)["size"]
+
+        structure = {
+            "nfreq": nfreq,
+            "ntime": ntime,
+            "metadata_size": msize,
+            "data_size": dsize,
+            "frame_size": _offset(
+                1 + msize + dsize, 4 * 1024
+            ),  # Align to 4kb boundaries
+        }
+
+        attributes = {
+            "git_version_tag": "hello",
+            "weight_type": "inverse_variance",
+            "instrument_name": "chime",
+        }
+
+        metadata = {
+            "index_map": index_map,
+            "structure": structure,
+            "attributes": attributes,
+        }
+
+        # Write metadata to file
+        with open(name + ".meta", "wb") as fh:
+            msgpack.dump(metadata, fh)
+
+        # Open the rawfile
+        rawfile = cls(name, mode="w+")
+
+        # Set the metadata on the frames that we already have
+        rawfile.valid_frames[:] = 1
+        rawfile.metadata["num_elements"][:] = ninput
+        rawfile.metadata["num_prod"][:] = nstack
+        rawfile.metadata["num_ev"][:] = nev
+        rawfile.metadata["freq_id"][:] = np.arange(nfreq)[np.newaxis, :]
+
+        fpga = np.array([t["fpga_count"] for t in time])
+        ctime = np.array(
+            [(int(np.floor(t["ctime"])), int((t["ctime"] % 1.0) * 1e9)) for t in time],
+            dtype=[("tv", np.int64), ("tv_nsec", np.uint64)],
+        )
+        rawfile.metadata["fpga_seq"][:] = fpga[:, np.newaxis]
+        rawfile.metadata["ctime"][:] = ctime[:, np.newaxis]
+
+        return rawfile
+
+    def flush(self):
+        """Ensure the data is flushed to disk."""
+        self.raw.flush()
+
+
+def simple_visraw_data(filename, ntime, nfreq, ninput):
+    """Create a simple VisRaw test file that kotekan can read.
+
+    Parameters
+    ----------
+    filename : str
+        Base name of files that will be written.
+    ninput, nfreq, ntime : int
+        Number of inputs, frequencies and times to use. These axes are given
+        dummy values.
+
+    Returns
+    -------
+    raw : VisRaw
+        A readonly view of the VisRaw file.
+    """
+
+    nprod = ninput * (ninput + 1) // 2
+    nev = 4
+
+    time = [{"ctime": (10.0 * i), "fpga_count": i} for i in range(ntime)]
+
+    freq = [{"centre": (800 - i * 10.0), "width": 10.0} for i in range(10)]
+
+    input_ = [(i, f"test{i:04}") for i in range(ninput)]
```

Comment:
This only works for python >= 3.7.
```suggestion
    input_ = [(i, "test{:04}".format(i)) for i in range(ninput)]
```